src/server/worker/api_server.go

```diff
@@ -1512,6 +1866,66 @@ func (a *APIServer) processDatums(pachClient *client.APIClient, logger *taggedLo
 	return result, nil
 }
 
+func (a *APIServer) setupStats(pachClient *client.APIClient, objClient obj.Client, tag string, stats *pps.ProcessStats, logger *taggedLogger, inputTree, outputTree *hashtree.Ordered, statsTree *hashtree.Unordered) func() error {
+	return func() error {
```

Comment:
It seems a little weird that this is returning a closure that does the work rather than just being a definition for the returned function. Callers can pretty easily turn it into a closure if they want to call it later, or just not call it until they're ready.

Also `setupStats` seems like a weird name for it, it seems like it's doing more than just setting them up, it seems like it's also merging them and writing them out.

Reply:
Yeah, that makes sense. This piece of code was changed a few times. It used to do more than create a function. The name setupStats was used to indicate that it was setting up the callback for writing stats. I will change this to a single function that will be deferred.